cli/registry.go

```diff
@@ -0,0 +1,254 @@
+package run
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/mitchellh/go-homedir"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+const defaultRegistryContainerName = "k3d-registry"
+
+const defaultRegistryImage = "registry:2"
+
+const defaultIntRegistryPorty = 5000
+
+const defaultFullRegistriesPath = "/etc/rancher/k3s/registries.yaml"
+
+var defaultRegistryLabels = map[string]string{
+	"app":       "k3d",
+	"component": "registry",
+}
+
+// NOTE: structs copied from https://github.com/rancher/k3s/blob/master/pkg/agent/templates/registry.go
+//       for avoiding a dependencies nightmare...
+
+// Registry is registry settings configured
+type Registry struct {
+	// Mirrors are namespace to mirror mapping for all namespaces.
+	Mirrors map[string]Mirror `toml:"mirrors" yaml:"mirrors"`
+
+	// Configs are configs for each registry.
+	// The key is the FDQN or IP of the registry.
+	Configs map[string]interface{} `toml:"configs" yaml:"configs"`
+
+	// Auths are registry endpoint to auth config mapping. The registry endpoint must
+	// be a valid url with host specified.
+	// DEPRECATED: Use Configs instead. Remove in containerd 1.4.
+	Auths map[string]interface{} `toml:"auths" yaml:"auths"`
+}
+
+// Mirror contains the config related to the registry mirror
+type Mirror struct {
+	// Endpoints are endpoints for a namespace. CRI plugin will try the endpoints
+	// one by one until a working one is found. The endpoint must be a valid url
+	// with host specified.
+	// The scheme, host and path from the endpoint URL will be used.
+	Endpoints []string `toml:"endpoint" yaml:"endpoint"`
+}
+
+// getGlobalRegistriesConfFilename gets the global registries file that will be used in all the servers/workers
+func getGlobalRegistriesConfFilename() (string, error) {
+	homeDir, err := homedir.Dir()
+	if err != nil {
+		log.Error("Couldn't get user's home directory")
+		return "", err
+	}
+
+	return path.Join(homeDir, ".k3d", "registries.yaml"), nil
+}
+
+// writeRegistriesConfigInContainer creates a valid registries configuration file in a container
+func writeRegistriesConfigInContainer(spec *ClusterSpec, ID string) error {
+	globalRegistriesConfig, err := getGlobalRegistriesConfFilename()
+	if err != nil {
+		return err
+	}
+	registryFullName := fmt.Sprintf("%s:%d", spec.RegistryName, spec.RegistryPort)
```

Comment:
Since we're now not changing the default internal port (from 5000), this will not work (i.e. it changes the target port for the k3d containers, but they won't reach the registry container on that port, because it's on default 5000 still) :+1:

Reply:
Good point. So we should use `defaultRegistryPort`...

Reply:
I've just realized we must make the distinction between internal and external addresses, right? For example, for `--registry-port 8888` we would need something like:

```yaml
mirrors:
  "registry.local:8888":
    endpoint:
    - http://registry.local:5000
```

Reply:
If you want to re-use the same full image name (which includes the registry name) for pulling (from within k3d/k3s) as used for pushing (from host system), then yes.

Reply:
As long as it's transparent to the user, I think it's ok :+1: